### ISSUE TITLE
parseMap rejects non-conflicting paths with overlapping prefix

### ISSUE THREAD

**Issue by joshng** [lightbend/config#644]
I have a test-framework that allows developers to override fragments of HOCON configuration to exercise config-driven behavior in test-cases. For convenience, my API allows config-paths to be specified as dotted strings, and supports overriding sub-objects by passing `Map<String, Object>` (actually, it performs JSON serialization of arbitrary Objects to convert them to `Maps`, but the integration with the `ConfigLoader` is the same):

```java
interface ConfigCustomizer {
   void overrideConfig(String path, Map<String, Object> value);
}
```

The paths/values provided via this interface are composed into a single `Map<String, Object>`, which is then parsed with `ConfigFactory.parseMap` to be used `withFallback`. This works nicely, but exhibits surprising behavior if multiple non-conflicting paths are specified that overlap at different levels of the hierarchy.

The documentation for `parseMap` suggests that keys which resolve to both "objects" and "values" will be rejected. However, the rejection occurs even if the paths are actually compatible and unambiguous (because the "value" is actually a nested, orthogonal Map).

To see the issue, first observe that a non-conflicting path which resolves to a nested `Map<String, Object>` is accepted by this method, and the paths in the nested Map are interpreted as you'd expect given the equivalent HOCON:
```HOCON
x {
  y.z: 1
}
```
... same as:
```java
Map<String, Object> map = new HashMap<>();
map.put("x", Collections.singletonMap("y.z", 1));
Config config = ConfigFactory.parseMap(map);
assertThat(config.getInt("x.y")).isEqualTo(1); // ok
```

However, if an additional path is added which is is unambiguously compatible, but which overlaps part of another path, an exception is thrown:
```
a.b.c: 0
a {
  d.e: 1
}
```

```java
Map<String, Object> map = new HashMap<>();
map.put("a.b.c", 0);
map.put("a", Collections.singletonMap("d.e", 1));  // overlaps with `a` above
Config config = ConfigFactory.parseMap(map);    // throws!
```

Here's the exception:
```
com.typesafe.config.ConfigException$BugOrBroken: In the map, path 'a' occurs as both the parent object of a value and as a value. Because Map has no defined ordering, this is a broken situation.
```

This restriction interferes with the goal to let developers freely specify related overrides in disparate areas of the code: if they use mixed styles to specify paths/values, confusion ensues.

FWIW, I can work around this issue by converting each individually-specified override into a separate `Config`, then composing them `withFallback` instead of putting them all into a single `Map` ... but the behavior still seems surprising and inconsistent (and the additional `Config`-parsing seems unnecessarily inefficient).

**Comment by havocp** [comment]
I guess the exception there is supposed to flag a case where an override would happen. (like both “a” contain “b.c”).  I agree it could be smarter, I don’t know how hard that is to implement though.

**Comment by joshng** [comment]
Yeah, I understood the situation the exception was meant to prevent, and agree that distinguishing real conflicts from valid compositions might be tricky (or maybe just inefficient, demanding a preliminary path-normalization step to grok the map contents). I'll review the implementation to consider how it might be made smarter, but if it's not worth fixing, then a doc and/or error-message update might be helpful to clarify the limitation.
